intermediate_source/parametrizations.py

```diff
@@ -0,0 +1,353 @@
+A = layer.weight
+print(A)
+assert torch.allclose(A, A.T)
+
+###############################################################################
+# We can do the same thing with any other layer. For example, we can create a CNN with
+# skew-symmetric kernels (:math:`X = -X^{\intercal}`). We use a similar parametrization,
+# copying minus the upper triangular part into the lower-triangular part
+class Skew(nn.Module):
+    def forward(self, X):
+        A = X.triu(1)
+        return A - A.transpose(-1, -2)
+
+
+cnn = nn.Conv2D(in_channels=5, out_channels=8, kernel_size=3)
+P.register_parametrization(layer, "weight", Skew())
+# Print a few kernels
+print(cnn.weight[0, 1])
+print(cnn.weight[2, 2])
+
+###############################################################################
+# Inspecting a parametrized module
+# --------------------------------
+# When a module is parametrized, we find that the module has changed a bit.
+# We may observe this by simply printing the module
+layer = nn.Linear(3, 3)
+print(f"Unparametrized:\n{layer}")
+P.register_parametrization(layer, "weight", Symmetric())
+print(f"Parametrized:\n{layer}")
+
+###############################################################################
+# We see that the ``Symmetric`` class has been registered under a ``parametrizations`` attribute.
+# This ``parametrizations`` attribute is an ``nn.ModuleList``, and it can be accessed as such
+print(layer.parametrizations.weight[0])
+
+###############################################################################
+# Note that each element in the `ModuleList` is itself a list, and we have to select the first
+# element of this list. It will be clear later the reason for this, when we see how to contactenate
+# paramtrizations.
+#
+# Something that we may notice is that, if we print the parameters, we see that the
+# parameter ``weight`` has been moved
+print(dict(layer.named_parameters()))
+
+###############################################################################
+# It now sits under ``layer.parametrizations.weight.original``
+print(layer.parametrizations.weight.original)
+
+###############################################################################
+# Besides these two small differences, the parametrization is doing exactly the same
+# as our manual implementation
+symmetric = Symmetric()
+assert torch.allclose(layer.weight, symmetric(layer.parametrizations.weight.original))
+
+###############################################################################
+# Parametrizations are first-class citizens
+# -----------------------------------------
+# Since ``layer.parametrizations`` is an `nn.ModuleList`, it means that the parametrizations
+# are properly registered as submodules of the original module. As such, the same rules
+# for registering parameters in a module apply to register a parametrization.
+# For example, if a parametrization has parameters, these will be moved from CPU
+# to CUDA when calling ``model = model.cuda()``.
```

Comment:
It may be good to add a note explaining how reparametrization magic is implemented under the hood

Reply:
I think that that would be too technical (it uses dynamically generated classes). I think something that could be done would be to stress more what happens after you call `register_parametrization`. In particular, the fact that it creates a `ModuleDict` under `module.parametrizations`, and each of those modules is a `ParametrizationList` and so on.
In fact, that's all `register_parametrization` is doing modulo the dynamically generated classes magic. This would also help clarifying why it can be used with `nn.Modules` but not with plain old functions.

Reply:
I think it's better to mention it, then there wouldn't be a surprise of some generated class name if there is some exception. Also this would be a useful reminder for the user to not do any similar magics if they fear interference

Reply:
I do agree that such explanation would be nice, but I'm not sure this tutorial is the right place for it.
A Note in the doc similar to the ones we have about autograd [herre](https://pytorch.org/docs/stable/notes/autograd.html) for example sound more appropriate.